src/security/intel/cbnt/measurement.c

```diff
@@ -0,0 +1,543 @@
+
+	/* Do nothing if BootGuard doesn't measure anything. */
+	if (!(biosacm_sts.status.bp & CBNT_BP_TYPE_M))
+		return;
+
+	/*
+	 * At this point the code should check whether PCRs 0-7 were capped with a digest of
+	 * value "1" (ASCII? binary? hex?) and if so, add appropriate entries to the event log.
+	 *
+	 * This can only happen if Bit3 of IbbFlags is set ("On error: Leave TPM Hierarchies
+	 * enabled. Cap all PCRs"), otherwise TPM is just disabled.  Either way Measured Boot
+	 * won't work, so skipping it unless reasons for doing it are discovered.
+	 */
+
+	/*
+	 * Making and hashing PCR-7 data.
+	 *
+	 * Pseudo-code of the data to be measured into PCR-0 for TigerLake and newer (older
+	 * hardware isn't supported):
+	 *
+	 *     struct {
+	 *          uint64_t ACM_POLICY_STATUS;
+	 *          uint16_t ACM.Header.SVN;
+	 *          uint8_t  ACM.Signature[ACM signature size];
+	 *          uint8_t  KM.Signature[KM signature size];
+	 *          uint8_t  BPM.Signature[BPM signature size];
+	 *          uint8_t  IBB.Digest[IBB digest size];
+	 *     } PCR0_DATA;
+	 */
+
+	/*
+	 * Allow for 4 8192-bit digests in PCR-0 measurements which is an unlikely situation,
+	 * so this buffer should be enough to not run out of space (data measured into PCR-7 is
+	 * smaller).
+	 */
+	char data[sizeof(uint64_t) + sizeof(uint16_t) + 4 * KiB];
+	struct obuf data_ob;
+	obuf_init(&data_ob, data, sizeof(data));
+
+	/* ACM_POLICY_STATUS (won't run out of space on this one) */
+	(void)obuf_write_le64(&data_ob, biosacm_sts.raw);
+
+	if (fill_pcr0_acm_fields(&data_ob) != CB_SUCCESS) {
+		printk(BIOS_ERR, "CBnT: failed to fill ACM fields of PCR-0 measurement data\n");
+		return;
+	}
+
+	if (fill_pcr0_km_fields(&data_ob) != CB_SUCCESS) {
+		printk(BIOS_ERR, "CBnT: failed to fill KM fields of PCR-0 measurement data\n");
+		return;
+	}
+
+	bool auth_measure;
+	if (fill_pcr0_bpm_fields(&data_ob, &auth_measure) != CB_SUCCESS) {
+		printk(BIOS_ERR, "CBnT: failed to fill BPM fields of PCR-0 measurement data\n");
+		return;
+	}
+
+	struct vb2_hash pcr0_hash;
+	if (vb2_hash_calculate(vboot_hwcrypto_allowed(), data, obuf_nr_written(&data_ob),
+			       tpm_log_alg(), &pcr0_hash)) {
+		printk(BIOS_ERR, "CBnT: failed to hash PCR-0 measurement data\n");
+		return;
+	}
+
+	/* Making and hashing PCR-7 data. */
+	struct vb2_hash pcr7_hash;
+	if (auth_measure) {
+		/* Reuse the first 2 fields of PCR-0 data which are identical in both cases. */
+		obuf_init(&data_ob, data, sizeof(data));
+		(void)obuf_oob_fill(&data_ob, sizeof(uint64_t) + sizeof(uint16_t));
+
+		if (!make_pcr7_hash(&data_ob, &pcr7_hash)) {
+			printk(BIOS_ERR,
+			       "CBnT: failed to build and hash PCR-7 measurement data\n");
+			return;
+		}
+	}
+
+	/*
+	 * BWG suggests to Perform reconstruction at this point, but we are likely to continue
+	 * the boot anyway as Measured Boot failures generally aren't fatal and incorrect hash
+	 * will be caught later during the boot process.  So not bothering.
+	 */
+
```

Comment:
I think we are also missing the `10.4.5.3 Startup Locality Event` https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClient_PFP_r1p05_v23_pub.pdf

if TPM is started from locality 3 instead of 0, then such information must be recorded in the event log.

```
The Startup Locality event should be placed in the log before any event which extends PCR[0]. This allows
software which needs to parse the TCG Event Log to initialize its internal PCR[0] state correctly. See Section
3.3.4.1 PCR[0] – SRTM, POST BIOS, and Embedded Drivers.
```

The startup locality is also controlled with a BPM flag and the locality used to perform startup is available in ACM_POLICY_STATUS bit 36

Reply:
Iimplemented, should work if whatever processes it allows event's data to be larger than expected: https://github.com/Dasharo/coreboot/pull/699/commits/24f501d37a12c56db9a38321402e32ca43e8a618 and https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-2191f194b15fa678d26fbe3aa4c3df04b08093f744b0964aca8e2270abcb7910.